src/protocol.rs

```diff
@@ -943,19 +948,68 @@ mod tests {
             .iter()
             .all(|p| *p.status() == Status::TerminatedSuccessfully));
 
+        // Set up keygen participants
+        let keygen_sid = Identifier::random(&mut rng);
+        let mut keygen_quorum = configs
+            .clone()
+            .into_iter()
+            .map(|config| {
+                Participant::<KeygenParticipant>::from_config(config, keygen_sid, ()).unwrap()
+            })
+            .collect::<Vec<_>>();
+        let mut keygen_outputs: HashMap<
+            ParticipantIdentifier,
+            <KeygenParticipant as ProtocolParticipant>::Output,
+        > = HashMap::new();
+
+        // Initialize keygen for all participants
+        for participant in &keygen_quorum {
+            let inbox = inboxes.get_mut(&participant.id).unwrap();
+            inbox.push(participant.initialize_message()?);
+        }
+
+        // Run keygen until all parties have outputs
+        while keygen_outputs.len() < QUORUM_SIZE {
+            let output = process_random_message(&mut keygen_quorum, &mut inboxes, &mut rng)?;
+
+            if let Some((pid, output)) = output {
+                // Save the output, and make sure this participant didn't already return an
+                // output.
+                assert!(keygen_outputs.insert(pid, output).is_none());
+            }
+        }
+
+        // Keygen is done! Makre sure there are no more messages.
+        assert!(inboxes_are_empty(&inboxes));
+        // And make sure all participants have successfully terminated.
+        assert!(keygen_quorum
+            .iter()
+            .all(|p| *p.status() == Status::TerminatedSuccessfully));
+
         // Tshare protocol
         // After the Tshare protocol, we will have a set of t-out-of-t shares
         // Therefore we need to remove some participants from the configs, and from
         // keygen and auxinfo outputs
+        let mut keygen_outputs_tshare = keygen_outputs.clone();
```

Comment:
I think it may be nice to have two separate test cases (like one full n out of n non-interactive and one t out of n on its own (no n out of n key gen setup). But I can see why you might want to do this to exercise all protocols at once.

Reply:
The unit tests called `basic_*` are the n out of n non-interactive. We have one for normal keys (using the keygen sub-protocol), and for HD wallet. We also have the threshold unit test, which exercises the t out of n case for both normal keys and HD wallet. So in total we have 4 cases being tested.